pkg/virtctl/credentials/add-key/add-key.go

```diff
@@ -0,0 +1,280 @@
+		Short: "Add credentials to a virtual machine.",
+		Args:  templates.ExactArgs("add-ssh-key", 1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runAddKeyCommand(clientConfig, cmdFlags, cmd, args)
+		},
+	}
+	cmdFlags.AddToCommand(cmd)
+
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+type addSshKeyFlags struct {
+	common.SshCommandFlags
+
+	CreateSecret bool
+	UpdateSecret bool
+
+	Force bool
+}
+
+func (a *addSshKeyFlags) AddToCommand(cmd *cobra.Command) {
+	a.SshCommandFlags.AddToCommand(cmd)
+
+	const (
+		createSecretFlag = "create-secret"
+		updateSecretFlag = "update-secret"
+	)
+
+	cmd.Flags().BoolVar(&a.CreateSecret, createSecretFlag, false, "Create a new secret for the SSH key. The new key will not be added to a running VM. Use --force to add a new secret even if the VM is running.")
+	cmd.Flags().BoolVar(&a.UpdateSecret, updateSecretFlag, false, "Add the SSH key to an existing secret. Use --force option, if the secret does not have owner reference pointing to the VM.")
+	cmd.MarkFlagsMutuallyExclusive(createSecretFlag, updateSecretFlag)
+
+	cmd.Flags().BoolVar(&a.Force, "force", false, "Force update of secret, even if it's not owned by the VM.")
+}
+
+func runAddKeyCommand(clientConfig clientcmd.ClientConfig, cmdFlags *addSshKeyFlags, cmd *cobra.Command, args []string) error {
+	vmName := args[0]
+	vmNamespace, _, err := clientConfig.Namespace()
+	if err != nil {
+		return fmt.Errorf("error getting namespace: %w", err)
+	}
+
+	// Reading the key before accessing cluster
+	sshKey, err := common.GetSshKey(&cmdFlags.SshCommandFlags)
+	if err != nil {
+		return fmt.Errorf("error getting ssh key: %w", err)
+	}
+
+	cli, err := kubecli.GetKubevirtClientFromClientConfig(clientConfig)
+	if err != nil {
+		return fmt.Errorf("error getting kubevirt client: %w", err)
+	}
+
+	vm, err := cli.VirtualMachine(vmNamespace).Get(cmd.Context(), vmName, &metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting virtual machine: %w", err)
+	}
+
+	if shouldCreateNewSecret(cmdFlags, vm) {
+		return addSecretWithSshKey(cmd, cli, cmdFlags, vm, sshKey)
+	} else {
+		return updateSecretWithSshKey(cmd, cli, cmdFlags, vm, sshKey)
+	}
+}
+
+func addSecretWithSshKey(cmd *cobra.Command, cli kubecli.KubevirtClient, cmdFlags *addSshKeyFlags, vm *v1.VirtualMachine, sshKey string) (err error) {
+	if !cmdFlags.Force {
+		// Only create a secret if VM is not running.
+		_, err := cli.VirtualMachineInstance(vm.Namespace).Get(cmd.Context(), vm.Name, &metav1.GetOptions{})
+		if err == nil {
+			return fmt.Errorf("virtual machine %s is running", vm.Name)
+		}
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("error when getting virtual machine instance: %w", err)
+		}
+	}
+
+	secret := newSecretWithKey(vm, sshKey)
+	secret, err = cli.CoreV1().Secrets(vm.Namespace).Create(cmd.Context(), secret, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating secret: %w", err)
+	}
+
+	vmPatched := false
```

Comment:
Do we really want to try to be clever and delete the secret again in case of an error? IMO it could could be kept and the code would be therefore simpler.

Reply:
done